### PR TITLE
Improve error reporting when fsfreeze fails

### PR DESCRIPTION
**Pull request by alromeros**
**What this PR does / why we need it**:

Fsfreeze can sometimes fail for causes unrelated to kubevirt, such as certain selinux policies or unsupported filesystems. This can lead to situations where snapshot fails without providing any pertinent information in any of the involved objects. This PR aims to add an event to avoid issues such as https://issues.redhat.com/browse/CNV-32664.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes # https://issues.redhat.com/browse/CNV-32664

**Special notes for your reviewer**:

Still working on adding a test but I'm really struggling to replicate a fsfreeze behavior, specially the selinux one described in the jira.

**Release note**:

```release-note
Bugfix: Improve error reporting when fsfreeze fails
```

### CODE REVIEW

**Comment by alromeros** [comment]
Attempted to replicate the SELinux error with a functional test but all our testing images currently disable SELinux. Since this is just an event and we don't provide any new functionality we can consider adding the test in the future.

File: tests/storage/snapshot.go [hunk 1]
```diff
@@ -733,6 +734,86 @@ var _ = SIGDescribe("VirtualMachineSnapshot Tests", func() {
...
+						},
+					},
+				})
+
+				vm, vmi = createAndStartVM(vm)
+				libwait.WaitForSuccessfulVMIStart(vmi,
+					libwait.WithTimeout(300),
+				)
+
+				Eventually(matcher.ThisVMI(vmi), 12*time.Minute, 2*time.Second).Should(matcher.HaveConditionTrue(v1.VirtualMachineInstanceAgentConnected))
```
[the remark this answers is not in the document]
Comment (alromeros) [reply tests/storage/snapshot.go:793]: Yeah this is a remnant from testing locally. Thanks for the review!

File: tests/storage/snapshot.go [hunk 2]
```diff
@@ -733,6 +734,70 @@ var _ = SIGDescribe("VirtualMachineSnapshot Tests", func() {
...
+				)
+
+				Eventually(matcher.ThisVMI(vmi), 1*time.Minute, 2*time.Second).Should(matcher.HaveConditionTrue(v1.VirtualMachineInstanceAgentConnected))
+				Expect(console.LoginToFedora(vmi)).To(Succeed())
+
+				// Recreating one specific SELinux error
+				Expect(console.SafeExpectBatch(vmi, []expect.Batcher{
+					&expect.BSnd{S: "mkdir /mount_dir\n"},
+					&expect.BExp{R: console.PromptExpression},
+					&expect.BSnd{S: "mkfs.ext4 /dev/vda1\n"},
```
Comment (akalenyu) [inline tests/storage/snapshot.go:787]: Can we always use vda1? I recall it backfired before with https://github.com/kubevirt/kubevirt/pull/10451

File: tests/storage/snapshot.go [hunk 3]
```diff
@@ -733,6 +734,70 @@ var _ = SIGDescribe("VirtualMachineSnapshot Tests", func() {
...
+					},
+				})
+				vm.Spec.Template.Spec.Volumes = append(vm.Spec.Template.Spec.Volumes, v1.Volume{
+					Name: "blank",
+					VolumeSource: v1.VolumeSource{
+						DataVolume: &v1.DataVolumeSource{
+							Name: dv.Name,
+						},
+					},
+				})
```
Comment (akalenyu) [inline tests/storage/snapshot.go:773]: libstorage.AddDataVolume may help

File: tests/storage/snapshot.go [hunk 4]
```diff
@@ -733,6 +734,70 @@ var _ = SIGDescribe("VirtualMachineSnapshot Tests", func() {
...
+
+				vm, vmi = createAndStartVM(vm)
+				libwait.WaitForSuccessfulVMIStart(vmi,
+					libwait.WithTimeout(300),
+				)
+
+				Eventually(matcher.ThisVMI(vmi), 1*time.Minute, 2*time.Second).Should(matcher.HaveConditionTrue(v1.VirtualMachineInstanceAgentConnected))
```
[the remark this answers is not in the document]
Comment (akalenyu) [reply tests/storage/snapshot.go:780]: Good point, maybe can just execute the sed command through console and then stop & start

File: tests/storage/snapshot.go [hunk 5]
```diff
@@ -733,6 +734,70 @@ var _ = SIGDescribe("VirtualMachineSnapshot Tests", func() {
...
+				Expect(console.LoginToFedora(vmi)).To(Succeed())
+
+				// Recreating one specific SELinux error
```
Comment (akalenyu) [inline tests/storage/snapshot.go:783]: Maybe we should add a reference to the bugzilla or some other readout here

File: tests/storage/snapshot.go (hunk 2 above)
Comment (alromeros) [reply tests/storage/snapshot.go:787]: Good catch.

**Review by 0xFelix — COMMENTED** [review]
One comment, otherwise looks good!

File: tests/storage/snapshot.go [hunk 6]
```diff
@@ -733,6 +734,71 @@ var _ = SIGDescribe("VirtualMachineSnapshot Tests", func() {
...
+				vm, vmi = createAndStartVM(vm)
+				libwait.WaitForSuccessfulVMIStart(vmi,
+					libwait.WithTimeout(300),
+				)
+
+				Eventually(matcher.ThisVMI(vmi), 1*time.Minute, 2*time.Second).Should(matcher.HaveConditionTrue(v1.VirtualMachineInstanceAgentConnected))
+
+				// Restart VM again to enable SELinux
+				Expect(virtClient.VirtualMachineInstance(vmi.Namespace).SoftReboot(context.Background(), vmi.Name)).ToNot(HaveOccurred())
+				Eventually(matcher.ThisVMI(vmi), 1*time.Minute, 2*time.Second).Should(matcher.HaveConditionTrue(v1.VirtualMachineInstanceAgentConnected))
```
[the remark this answers is not in the document]
Comment (alromeros) [reply tests/storage/snapshot.go:767]: The softreboot tests usually wait 30 secs after a reboot, but I can extend it here (maybe up to 3 mins?).

File: tests/storage/snapshot.go [hunk 7]
```diff
@@ -733,6 +734,71 @@ var _ = SIGDescribe("VirtualMachineSnapshot Tests", func() {
...
+					&expect.BSnd{S: fmt.Sprintf("mount %s /mount_dir\n", blankDisk)},
+					&expect.BExp{R: console.PromptExpression},
+				}, 20)).To(Succeed())
+
+				snapshot = newSnapshot()
+				_, err = virtClient.VirtualMachineSnapshot(snapshot.Namespace).Create(context.Background(), snapshot, metav1.CreateOptions{})
+				Expect(err).ToNot(HaveOccurred())
+
+				objectEventWatcher := watcher.New(vmi).SinceWatchedObjectResourceVersion().Timeout(time.Duration(30) * time.Second)
+				objectEventWatcher.WaitFor(context.Background(), watcher.WarningEvent, "FreezeError")
```
Comment (0xFelix) [inline tests/storage/snapshot.go:799]: Is the result of this asserted/expected somehow? How do we know it did not timeout?

Comment (alromeros) [reply tests/storage/snapshot.go:799]: Do you mean the `objectEventWatcher.WaitFor(context.Background(), watcher.WarningEvent, "FreezeError")`? It fails after timeout if the event is not found.

**Review by 0xFelix — APPROVED** [review]
/lgtm

File: tests/storage/snapshot.go [hunk 8]
```diff
@@ -733,6 +734,71 @@ var _ = SIGDescribe("VirtualMachineSnapshot Tests", func() {
 				}
 			})
 
+			It("should report appropriate event when freeze fails", func() {
```
Comment (xpivarc) [inline tests/storage/snapshot.go:737]: @mhenriks this could have been a unit test.

Comment (mhenriks) [reply tests/storage/snapshot.go:737]: Yeah my concern was that there are currently no unit tests in that package

**Pull request merged**